### PR TITLE
Remove LocalCheckpointTracker#resetCheckpoint

### PR DESCRIPTION
**Pull request by dnhatn**
In #34474, we added a [new assertion](https://github.com/elastic/elasticsearch/blob/ab5a0b15f6205ed426895e2805d8e4fb56fdd657/server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java#L685-L686) to ensure that the LocalCheckpointTracker is always consistent with Lucene index. However, we reset LocalCheckpoinTracker in testDedupByPrimaryTerm cause this assertion to be [violated](
https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+6.x+multijob-unix-compatibility/os=ubuntu&&virtual/8/console).

This commit removes resetCheckpoint from LocalCheckpointTracker and rewrites testDedupByPrimaryTerm without resetting the local checkpoint.

### CODE REVIEW

**Comment by bleskes** [comment]
@dnhatn does that test even have value today? I think that we now rollback, we should never override a term in lucene? (actually we should never override an operation)

**Comment by dnhatn** [comment]
@bleskes I renamed and added a comment for the test. It's ready for another round.

File: server/src/test/java/org/elasticsearch/index/engine/LuceneChangesSnapshotTests.java [hunk 1]
```diff
@@ -150,35 +150,76 @@ public void testBasics() throws Exception {
...
-                }
-                if (rarely()) {
-                    engine.flush();
+                    totalOps += ((Engine.Index) op).docs().size();
+                } else {
+                    totalOps++;
                 }
-                totalOps++;
+            }
+            if (op instanceof Engine.Index) {
```
Comment (bleskes) [inline server/src/test/java/org/elasticsearch/index/engine/LuceneChangesSnapshotTests.java]: we do this so often. I wonder if it's time for a utility method.

File: server/src/test/java/org/elasticsearch/index/engine/LuceneChangesSnapshotTests.java [hunk 2]
```diff
@@ -150,35 +150,76 @@ public void testBasics() throws Exception {
...
+                        throw new IllegalStateException("invalid operation type [" + opType + "]");
                 }
+                operations.add(op);
+            }
+        }
+        Randomness.shuffle(operations);
+        int totalOps = 0;
+        for (Engine.Operation op : operations) {
+            // Engine skips deletes or indexes below the local checkpoint
+            if (engine.getLocalCheckpoint() < op.seqNo() || op instanceof Engine.NoOp) {
```
Comment (bleskes) [inline server/src/test/java/org/elasticsearch/index/engine/LuceneChangesSnapshotTests.java]: The fact that we process noops differently than indexing / delete ops (w.r.t localcheckpoint) sounds like a bug (different) PR)

**Review by bleskes — APPROVED** [review]
LGTM

**Comment by bleskes** [comment]
> We can make the Engine to index once. WDYT?

I think this is a good thing to do and we discussed it in the context of the following engine. I think that to fully do it we need to have soft deletes and restore the local checkpoint tracker correctly when opening the engine. Based on that, I think we need to wait until soft deletes become a default.

File: server/src/test/java/org/elasticsearch/index/engine/LuceneChangesSnapshotTests.java (hunk 1 above)
Comment (dnhatn) [reply server/src/test/java/org/elasticsearch/index/engine/LuceneChangesSnapshotTests.java]: Yes, I added an util :).

File: server/src/test/java/org/elasticsearch/index/engine/LuceneChangesSnapshotTests.java (hunk 2 above)
Comment (dnhatn) [reply server/src/test/java/org/elasticsearch/index/engine/LuceneChangesSnapshotTests.java]: sure, I will make this in a follow up.

**Comment by dnhatn** [comment]
Thanks @bleskes.

**Pull request merged**